addons/stock/models/stock_lot.py

```diff
@@ -291,3 +292,70 @@ def _find_delivery_ids_by_lot(self, lot_path=None, delivery_by_lot=None):
 
             delivery_by_lot[lot.id] = list(delivery_ids)
         return delivery_by_lot
+
+    def _find_delivery_ids_by_lot_iterative(self):
+        """ Retrieve all delivery IDs (outgoing picking) linked to the lots
+            in self and all the lots found when parcouring the produce lines.
+            :return: A dictionary where keys are the IDs of the original 'stock.lot'
+                      records (self) and values are lists of associated 'stock.picking' IDs.
+            :rtype: dict
+        """
+
+        all_lot_ids = set(self.ids)
+        barren_lines = defaultdict(set)
+        parent_map = defaultdict(set)
+
+        # Prefetch the lines linked to lots and split them between producing lines
+        # and barren lines (lines that have `produce_line_ids` and lines that don't
+        # have them respectively) and build the map of the parents of each lot (so we
+        # can browse the tree from the leaves to the root and propagate the pickings)
+        queue = list(self.ids)
+        while queue:
+            domain = [
+                ('lot_id', 'in', queue),
+                ('state', '=', 'done'),
+            ]
+            domain_restriction = self._get_outgoing_domain()
+            domain = expression.AND([domain, domain_restriction])
+
+            queue = []
+            move_lines = self.env['stock.move.line'].search(domain)
+            for line in move_lines:
+                lot_id = line.lot_id.id
+
+                produce_line_lot_ids = line.produce_line_ids.lot_id.ids
+                if produce_line_lot_ids:
+                    for child_lot_id in produce_line_lot_ids:
+                        parent_map[child_lot_id].add(lot_id)
+                else:
+                    barren_lines[lot_id].add(line.id)
+
+                next_lots = set(produce_line_lot_ids) - all_lot_ids
+                all_lot_ids.update(next_lots)
+                queue.extend(next_lots)
+
+        # Initialize delivery_by_lot with barren lines (i.e. the leaves of the lot tree)
+        lots_to_propagate = set()
+        delivery_by_lot = {lot_id: set() for lot_id in all_lot_ids}
+        for lot_id in all_lot_ids:
```

Comment:
```suggestion
        for lot_id in barren_lines.keys():
```
If we want to save some *minor* time, but doesn't make sense to loop over other lots.